File: lib/java/src/protocol/TBinaryProtocol.java
package com.facebook.thrift.protocol;

import com.facebook.thrift.TException;
import com.facebook.thrift.transport.TTransport;

/**
 * Binary protocol implementation for thrift.
 *
 * @author Mark Slee <mcslee@facebook.com>
 */
public class TBinaryProtocol implements TProtocol {

  public void writeMessageBegin(TTransport out, TMessage message) throws TException {
    writeString(out, message.name);
    writeByte(out, message.type);
    writeI32(out, message.seqid);
  }

  public void writeMessageEnd(TTransport out) throws TException {}


  public void writeStructBegin(TTransport out, TStruct struct) throws TException {}

  public void writeStructEnd(TTransport out) throws TException {}

  public void writeFieldBegin(TTransport out, TField field) throws TException {
    writeByte(out, field.type);
    writeI16(out, field.id);
  }

  public void writeFieldEnd(TTransport out) throws TException {}

  public void writeFieldStop(TTransport out) throws TException {
    writeByte(out, TType.STOP);
  }

  public void writeMapBegin(TTransport out, TMap map) throws TException {
    writeByte(out, map.keyType);
    writeByte(out, map.valueType);
    writeI32(out, map.size);
  }

  public void writeMapEnd(TTransport out) throws TException {}

  public void writeListBegin(TTransport out, TList list) throws TException {
    writeByte(out, list.elemType);
    writeI32(out, list.size);
  }

  public void writeListEnd(TTransport out) throws TException {}

  public void writeSetBegin(TTransport out, TSet set) throws TException {
    writeByte(out, set.elemType);
    writeI32(out, set.size);
  }

  public void writeSetEnd(TTransport out) throws TException {}

  public void writeBool(TTransport out, boolean b) throws TException {
    writeByte(out, b ? (byte)1 : (byte)0);
  }

  byte[] bout = new byte[1];
  public void writeByte(TTransport out, byte b) throws TException {
    bout[0] = b;
    out.write(bout, 0, 1);
  }

  byte[] i16out = new byte[2];
  public void writeI16(TTransport out, short i16) throws TException {
    i16out[0] = (byte)(0xff & (i16 >> 8));
    i16out[1] = (byte)(0xff & (i16));
    out.write(i16out, 0, 2);
  }

  byte[] i32out = new byte[4];
  public void writeI32(TTransport out, int i32) throws TException {
    i32out[0] = (byte)(0xff & (i32 >> 24));
    i32out[1] = (byte)(0xff & (i32 >> 16));
    i32out[2] = (byte)(0xff & (i32 >> 8));
    i32out[3] = (byte)(0xff & (i32));
    out.write(i32out, 0, 4);
  }

  byte[] i64out = new byte[8];
  public void writeI64(TTransport out, long i64) throws TException {
    i64out[0] = (byte)(0xff & (i64 >> 56));
    i64out[1] = (byte)(0xff & (i64 >> 48));
    i64out[2] = (byte)(0xff & (i64 >> 40));
    i64out[3] = (byte)(0xff & (i64 >> 32));
    i64out[4] = (byte)(0xff & (i64 >> 24));
    i64out[5] = (byte)(0xff & (i64 >> 16));
    i64out[6] = (byte)(0xff & (i64 >> 8));
    i64out[7] = (byte)(0xff & (i64));
    out.write(i64out, 0, 8);
  }

  public void writeDouble(TTransport out, double dub) throws TException {
    writeI64(out, Double.doubleToLongBits(dub));
  }

  public void writeString(TTransport out, String str) throws TException {
    byte[] dat = str.getBytes();
    writeI32(out, dat.length);
    out.write(dat, 0, dat.length);
  }

  /**
   * Reading methods.
   */

  public TMessage readMessageBegin(TTransport in) throws TException {
    TMessage message = new TMessage();
    message.name = readString(in);
    message.type = readByte(in);
    message.seqid = readI32(in);
    return message;
  }

  public void readMessageEnd(TTransport in) throws TException {}

  public TStruct readStructBegin(TTransport in) throws TException {
    return new TStruct();
  }

  public void readStructEnd(TTransport in) throws TException {}

  public TField readFieldBegin(TTransport in) throws TException {
    TField field = new TField();
    field.type = readByte(in);
    if (field.type != TType.STOP) {
      field.id = readI16(in);
    }
    return field;
  }
  
  public void readFieldEnd(TTransport in) throws TException {}
 
  public TMap readMapBegin(TTransport in) throws TException {
    TMap map = new TMap();
    map.keyType = readByte(in);
    map.valueType = readByte(in);
    map.size = readI32(in);
    return map;
  }

  public void readMapEnd(TTransport in) throws TException {}

  public TList readListBegin(TTransport in) throws TException {
    TList list = new TList();
    list.elemType = readByte(in);
    list.size = readI32(in);
    return list;
  }

  public void readListEnd(TTransport in) throws TException {}

  public TSet readSetBegin(TTransport in) throws TException {
    TSet set = new TSet();
    set.elemType = readByte(in);
    set.size = readI32(in);
    return set;
  }

  public void readSetEnd(TTransport in) throws TException {}

  public boolean readBool(TTransport in) throws TException {
    return (readByte(in) == 1);
  }

  byte[] bin = new byte[1];
  public byte readByte(TTransport in) throws TException {
    in.readAll(bin, 0, 1);
    return bin[0];
  }

  byte[] i16rd = new byte[2];
  public short readI16(TTransport in) throws TException {
    in.readAll(i16rd, 0, 2);
    return
      (short)
      (((i16rd[0] & 0xff) << 8) |
       ((i16rd[1] & 0xff)));
  }

  byte[] i32rd = new byte[4];
  public int readI32(TTransport in) throws TException {
    in.readAll(i32rd, 0, 4);
    return
      ((i32rd[0] & 0xff) << 24) |
      ((i32rd[1] & 0xff) << 16) |
      ((i32rd[2] & 0xff) <<  8) |
      ((i32rd[3] & 0xff));
  }
 
  byte[] i64rd = new byte[8];
  public long readI64(TTransport in) throws TException {
    in.readAll(i64rd, 0, 8);
    return
      ((long)(i64rd[0] & 0xff) << 56) |
      ((long)(i64rd[1] & 0xff) << 48) |
      ((long)(i64rd[2] & 0xff) << 40) |
      ((long)(i64rd[3] & 0xff) << 32) |
      ((long)(i64rd[4] & 0xff) << 24) |
      ((long)(i64rd[5] & 0xff) << 16) |
      ((long)(i64rd[6] & 0xff) <<  8) |
      ((long)(i64rd[7] & 0xff));
  }

  public double readDouble(TTransport in) throws TException {
    return Double.longBitsToDouble(readI64(in));
  }

  public String readString(TTransport in)  throws TException {
    int size = readI32(in);
    byte[] buf = new byte[size];
    in.readAll(buf, 0, size);
    return new String(buf);
  }
}

File: lib/dart/lib/src/protocol/t_binary_protocol.dart
/// Licensed to the Apache Software Foundation (ASF) under one
/// or more contributor license agreements. See the NOTICE file
/// distributed with this work for additional information
/// regarding copyright ownership. The ASF licenses this file
/// to you under the Apache License, Version 2.0 (the
/// "License"); you may not use this file except in compliance
/// with the License. You may obtain a copy of the License at
///
/// http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing,
/// software distributed under the License is distributed on an
/// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
/// KIND, either express or implied. See the License for the
/// specific language governing permissions and limitations
/// under the License.

part of thrift;

class TBinaryProtocolFactory implements TProtocolFactory<TBinaryProtocol> {
  TBinaryProtocolFactory({this.strictRead = false, this.strictWrite = true});

  final bool strictRead;
  final bool strictWrite;

  @override
  TBinaryProtocol getProtocol(TTransport transport) {
    return TBinaryProtocol(transport,
        strictRead: strictRead, strictWrite: strictWrite);
  }
}

/// Binary protocol implementation for Thrift.
///
/// Adapted from the C# version.
class TBinaryProtocol extends TProtocol {
  static const int VERSION_MASK = 0xffff0000;
  static const int VERSION_1 = 0x80010000;

  static const Utf8Codec _utf8Codec = Utf8Codec();

  final bool strictRead;
  final bool strictWrite;

  TBinaryProtocol(TTransport transport,
      {this.strictRead = false, this.strictWrite = true})
      : super(transport);

  /// write
  @override
  void writeMessageBegin(TMessage message) {
    if (strictWrite) {
      int version = VERSION_1 | message.type;
      writeI32(version);
      writeString(message.name);
      writeI32(message.seqid);
    } else {
      writeString(message.name);
      writeByte(message.type);
      writeI32(message.seqid);
    }
  }

  @override
  void writeMessageEnd() {}

  @override
  void writeStructBegin(TStruct struct) {}

  @override
  void writeStructEnd() {}

  @override
  void writeFieldBegin(TField field) {
    writeByte(field.type);
    writeI16(field.id);
  }

  @override
  void writeFieldEnd() {}

  @override
  void writeFieldStop() {
    writeByte(TType.STOP);
  }

  @override
  void writeMapBegin(TMap map) {
    writeByte(map.keyType);
    writeByte(map.valueType);
    writeI32(map.length);
  }

  @override
  void writeMapEnd() {}

  @override
  void writeListBegin(TList list) {
    writeByte(list.elementType);
    writeI32(list.length);
  }

  @override
  void writeListEnd() {}

  @override
  void writeSetBegin(TSet set) {
    writeByte(set.elementType);
    writeI32(set.length);
  }

  @override
  void writeSetEnd() {}

  @override
  void writeBool(bool b) {
    if (b == null) b = false;
    writeByte(b ? 1 : 0);
  }

  final ByteData _byteOut = ByteData(1);

  @override
  void writeByte(int byte) {
    if (byte == null) byte = 0;
    _byteOut.setUint8(0, byte);
    transport.write(_byteOut.buffer.asUint8List(), 0, 1);
  }

  final ByteData _i16Out = ByteData(2);

  @override
  void writeI16(int i16) {
    if (i16 == null) i16 = 0;
    _i16Out.setInt16(0, i16);
    transport.write(_i16Out.buffer.asUint8List(), 0, 2);
  }

  final ByteData _i32Out = ByteData(4);

  @override
  void writeI32(int i32) {
    if (i32 == null) i32 = 0;
    _i32Out.setInt32(0, i32);
    transport.write(_i32Out.buffer.asUint8List(), 0, 4);
  }

  final Uint8List _i64Out = Uint8List(8);

  @override
  void writeI64(int i64) {
    if (i64 == null) i64 = 0;
    var i = Int64(i64);
    var bts = i.toBytes();
    for (var j = 0; j < 8; j++) {
      _i64Out[j] = bts[8 - j - 1];
    }
    transport.write(_i64Out, 0, 8);
  }

  @override
  void writeString(String s) {
    var bytes = s != null ? _utf8Codec.encode(s) : Uint8List.fromList([]);
    writeI32(bytes.length);
    transport.write(bytes, 0, bytes.length);
  }

  final ByteData _doubleOut = ByteData(8);

  @override
  void writeDouble(double d) {
    if (d == null) d = 0.0;
    _doubleOut.setFloat64(0, d);
    transport.write(_doubleOut.buffer.asUint8List(), 0, 8);
  }

  @override
  void writeBinary(Uint8List bytes) {
    var length = bytes.length;
    writeI32(length);
    transport.write(bytes, 0, length);
  }

  /// read
  @override
  TMessage readMessageBegin() {
    String name;
    int type;
    int seqid;

    int size = readI32();
    if (size < 0) {
      int version = size & VERSION_MASK;
      if (version != VERSION_1) {
        throw TProtocolError(TProtocolErrorType.BAD_VERSION,
            "Bad version in readMessageBegin: $version");
      }
      type = size & 0x000000ff;
      name = readString();
      seqid = readI32();
    } else {
      if (strictRead) {
        throw TProtocolError(TProtocolErrorType.BAD_VERSION,
            "Missing version in readMessageBegin");
      }
      name = _readString(size);
      type = readByte();
      seqid = readI32();
    }
    return TMessage(name, type, seqid);
  }

  @override
  void readMessageEnd() {}

  @override
  TStruct readStructBegin() {
    return TStruct();
  }

  @override
  void readStructEnd() {}

  @override
  TField readFieldBegin() {
    String name = "";
    int type = readByte();
    int id = type != TType.STOP ? readI16() : 0;

    return TField(name, type, id);
  }

  @override
  void readFieldEnd() {}

  @override
  TMap readMapBegin() {
    int keyType = readByte();
    int valueType = readByte();
    int length = readI32();

    return TMap(keyType, valueType, length);
  }

  @override
  void readMapEnd() {}

  @override
  TList readListBegin() {
    int elementType = readByte();
    int length = readI32();

    return TList(elementType, length);
  }

  @override
  void readListEnd() {}

  @override
  TSet readSetBegin() {
    int elementType = readByte();
    int length = readI32();

    return TSet(elementType, length);
  }

  @override
  void readSetEnd() {}

  @override
  bool readBool() => readByte() == 1;

  final Uint8List _byteIn = Uint8List(1);

  @override
  int readByte() {
    transport.readAll(_byteIn, 0, 1);
    return _byteIn.buffer.asByteData().getUint8(0);
  }

  final Uint8List _i16In = Uint8List(2);

  @override
  int readI16() {
    transport.readAll(_i16In, 0, 2);
    return _i16In.buffer.asByteData().getInt16(0);
  }

  final Uint8List _i32In = Uint8List(4);

  @override
  int readI32() {
    transport.readAll(_i32In, 0, 4);
    return _i32In.buffer.asByteData().getInt32(0);
  }

  final Uint8List _i64In = Uint8List(8);

  @override
  int readI64() {
    transport.readAll(_i64In, 0, 8);
    var i = Int64.fromBytesBigEndian(_i64In);
    return i.toInt();
  }

  final Uint8List _doubleIn = Uint8List(8);

  @override
  double readDouble() {
    transport.readAll(_doubleIn, 0, 8);
    return _doubleIn.buffer.asByteData().getFloat64(0);
  }

  @override
  String readString() {
    int size = readI32();
    return _readString(size);
  }

  String _readString(int size) {
    Uint8List stringIn = Uint8List(size);
    transport.readAll(stringIn, 0, size);
    return _utf8Codec.decode(stringIn);
  }

  @override
  Uint8List readBinary() {
    int length = readI32();
    Uint8List binaryIn = Uint8List(length);
    transport.readAll(binaryIn, 0, length);
    return binaryIn;
  }
}
